### PR TITLE
Enforce eval shuffle warning only for default samplers

### LINKED ISSUE #11856: Even when val DataLoader has shuffle=False, Lightning gives an incorrect warning that val Dataloader has shuffle=True
## 🐛 Bug
Even when DataLoader(self.valid_data, shuffle=False, ...), Lightning gives the following message:   UserWarning: Your `val_dataloader` has `shuffle=True`,.......

### To Reproduce
Assign shuffle=False in both the val and test dataloaders. Run the code and see the following two warning messages:

> .../.local/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py:659: UserWarning: Your `val_dataloader` has `shuffle=True`, it is strongly recommended that you turn this off for val/test/predict dataloaders.

> .local/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py:659: UserWarning: Your `test_dataloader` has `shuffle=True`, it is strongly recommended that you turn this off for val/test/predict dataloaders.

### Expected behavior
There should be no warning messages.

### Environment
- PyTorch Lightning Version (e.g., 1.5.0): 1.5.9
- PyTorch Version (e.g., 1.10): 1.10.2
- Python version (e.g., 3.9): 3.8.10
- OS (e.g., Linux): Linux (Ubuntu)
- CUDA/cuDNN version: cu113
- GPU models and configuration: GTX 1080
- How you installed PyTorch (`conda`, `pip`, source): pip

### Additional context
Previous versions of Lightning also gave those warning messages

cc @justusschock @awaelchli @ninginthecloud @tchaton @rohitgr7 @akihironitta

### LINKED ISSUE #12648: ```_check_eval_shuffling``` issues warning when sampler is a NumPy array
## 🐛 Bug

The function ```pytorch_lightning.trainer.data_loading._check_eval_shuffling``` issues a warning if the sampler of a Map-style Dataset is defined and is not a SequentialSampler.
The warning however, says that the issue is that the val_dataset has ```shuffle=True```, which is not necessarily true.
For instance, if you use a simple NumPy array as sampler, the warning is issued but the data loader is completely deterministic.

### To Reproduce

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2, sampler=list(range(64)))
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()
```

### Expected behavior

I see three possible solutions:
* The warning is not issued if the sampler is an Iterable whose order is fixed.
* The warning is clearer in stating what is triggering it.
* The warning specifies it can be ignored if you are using a list/array as sampler.

### Environment

- PyTorch Lightning Version: 1.5.9
- PyTorch Version: 1.10.2
- Python version: 3.8.3
- OS: MacOS
- How you installed PyTorch: poetry

cc @justusschock @awaelchli @ninginthecloud @rohitgr7 @carmocca @borda @ananthsub @jjenniferdai

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12648
Fixes #11856

For reviewers: https://github.com/PyTorchLightning/pytorch-lightning/commit/a179fdce4117ad39af5f8cd0edc809c8d39198cc is the only commit that needs review. Rest is just moving tests from test_data_loading to test_data_connector since we have deprecated data_loading. Moved them since it becomes hard to find tests that belong to data_connector.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @ninginthecloud @rohitgr7

### CODE REVIEW

File: tests/helpers/utils.py [hunk 1]
```diff
@@ -126,7 +127,7 @@ def no_warning_call(expected_warning: Type[Warning] = UserWarning, match: Option
             return
     else:
         for w in record.list:
-            if w.category is expected_warning and match in w.message.args[0]:
+            if w.category is expected_warning and re.compile(match).search(str(w.message.args[0])):
```
Comment (rohitgr7) [inline tests/helpers/utils.py:130]: this is a small fix to ensure we can match regex patterns.

**Comment by rohitgr7** [comment]
for reviewers: a179fdce4117ad39af5f8cd0edc809c8d39198cc is the only commit that needs review. Rest is just moving tests from `test_data_loading` to `test_data_connector` since we have deprecated `data_loading`. Moved them since it becomes hard to find tests that belong to `data_connector`.

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 2]
```diff
@@ -460,7 +460,12 @@ def replace_sampler(dataloader: DataLoader) -> DataLoader:
 
     @staticmethod
     def _check_eval_shuffling(dataloader, mode):
-        if _is_dataloader_shuffled(dataloader):
+        # limit this warning only for samplers assigned automatically when shuffle is set
+        if (
+            hasattr(dataloader, "sampler")
+            and isinstance(dataloader.sampler, (RandomSampler, SequentialSampler))
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/data_connector.py:466]: `is_dataloader_shuffled` is a very loose check due to the behavior of how `shuffle` is treated in the `DataLoader`. I am trying to create a `we don't care situation` here if it's a custom sampler.

If we keep it inside `is_dataloader_shuffled` then it's hard to predict whether the sampler is working sequentially or not if it's a custom one. Right now it returns True if the sampler is not a sequential sampler, but if we move it there, then we need to return False to satisfy the condition here.

File: tests/helpers/utils.py (hunk 1 above)
Comment (rohitgr7) [inline tests/helpers/utils.py:130]:
```suggestion
            if w.category is expected_warning and re.compile(match).search(w.message.args[0]):
```

File: tests/trainer/connectors/test_data_connector.py [hunk 3]
```diff
@@ -11,20 +11,388 @@
...
-from tests.helpers import BoringDataModule, BoringModel
-from tests.helpers.boring_model import RandomDataset
+from tests.helpers.boring_model import BoringDataModule, BoringModel, RandomDataset
+from tests.helpers.runif import RunIf
 from tests.helpers.utils import no_warning_call
 
 
+@RunIf(skip_windows=True)
+@pytest.mark.parametrize("mode", (1, 2))
+def test_replace_distributed_sampler(tmpdir, mode):
```
Comment (awaelchli) [inline tests/trainer/connectors/test_data_connector.py:37]: Can you please point out which tests are new? Or push refactor to a new PR. It makes a reviewers life much easier :)

Comment (rohitgr7) [reply tests/trainer/connectors/test_data_connector.py:37]: mentioned here: https://github.com/PyTorchLightning/pytorch-lightning/pull/12653#issuecomment-1091595756

Comment (awaelchli) [reply tests/trainer/connectors/test_data_connector.py:37]: added to description.

Comment (rohitgr7) [reply tests/trainer/connectors/test_data_connector.py:37]: just came back here to do that... ahaaha! better :) thanks

File: tests/trainer/connectors/test_data_connector.py [hunk 4]
```diff
@@ -11,20 +11,388 @@
...
+    trainer._data_connector.attach_data(model)
+    with pytest.raises(
+        MisconfigurationException,
+        match=rf"{limit_val_batches} \* {dl_size} < 1. Please increase the `limit_val_batches`",
+    ):
+        trainer._data_connector._reset_eval_dataloader(RunningStage.VALIDATING, model)
+
+
+@pytest.mark.parametrize(
+    "val_dl,raises_exception",
```
Comment (awaelchli) [inline tests/trainer/connectors/test_data_connector.py:365]:
```suggestion
    "val_dl,warns",
```

**Review by awaelchli — APPROVED** [review]

**Review by krshrimali — APPROVED** [review]
Thanks! LGTM

File: tests/trainer/connectors/test_data_connector.py [hunk 5]
```diff
@@ -11,20 +11,388 @@
...
+            ),
+            True,
+        ),
+    ],
+)
+def test_non_sequential_sampler_warning_is_raised_for_eval_dataloader(val_dl, raises_exception):
+    trainer = Trainer()
+    model = BoringModel()
+    trainer._data_connector.attach_data(model, val_dataloaders=val_dl)
+    context = pytest.warns if raises_exception else no_warning_call
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_data_connector.py:391]:
```suggestion
def test_non_sequential_sampler_warning_is_raised_for_eval_dataloader(val_dl, warns):
    trainer = Trainer()
    model = BoringModel()
    trainer._data_connector.attach_data(model, val_dataloaders=val_dl)
    context = pytest.warns if warns else no_warning_call
```

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/data_connector.py]: I did not completely understand your argument here. However, I tried implementing my idea and it seems to be passing the tests added: https://github.com/PyTorchLightning/pytorch-lightning/pull/12653/commits/a030de37e303f8e4ffd973da8960e494fda83497

Can you double check it? Do we need to add a test case for an edge I missed?

Thank you!

**Review by carmocca — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
